File: android/src/main/kotlin/dev/depthlift/DepthMesh.kt
package dev.depthlift

import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.FloatBuffer
import java.nio.IntBuffer

/**
 * Builds a planar triangle-strip mesh displaced by a depth map.
 *
 * The mesh is a regular grid of [resolution] × [resolution] quads,
 * each split into two triangles. Vertex positions are in normalised
 * device coordinates [-1, 1], UVs are [0, 1], and per-vertex depth
 * values are read from the supplied depth map bytes.
 *
 * Foreground vertices (depth > [FOREGROUND_THRESHOLD]) receive 1.5×
 * amplified Z displacement for a pop-out feel.
 *
 * @param resolution number of subdivisions along each axis (16–256).
 * @param imageWidth  original image width in pixels.
 * @param imageHeight original image height in pixels.
 * @param depthBytes  raw depth map — expected as 16-bit unsigned values,
 *                    two bytes per pixel, little-endian.
 */
class DepthMesh(
    private val resolution: Int,
    private val imageWidth: Int,
    private val imageHeight: Int,
    private val depthBytes: ByteArray
) {
    /** XY positions in NDC, 2 floats per vertex. */
    val positions: FloatBuffer

    /** UV texture coordinates, 2 floats per vertex. */
    val uvs: FloatBuffer

    /** Normalised depth value, 1 float per vertex. */
    val depths: FloatBuffer

    /** Triangle indices (unsigned int). */
    val indices: IntBuffer

    /** Total number of vertices in the mesh. */
    val vertexCount: Int

    /** Total number of index elements. */
    val indexCount: Int

    init {
        val cols = resolution + 1
        val rows = resolution + 1
        vertexCount = cols * rows

        // Two triangles per quad, 3 indices each.
        indexCount = resolution * resolution * 6

        positions = allocFloat(vertexCount * 2)
        uvs = allocFloat(vertexCount * 2)
        depths = allocFloat(vertexCount)
        indices = allocInt(indexCount)

        buildVertices(cols, rows)
        buildIndices(cols, rows)

        positions.position(0)
        uvs.position(0)
        depths.position(0)
        indices.position(0)
    }

    // ─── Vertex generation ────────────────────────────────────────────────

    private fun buildVertices(cols: Int, rows: Int) {
        for (row in 0 until rows) {
            for (col in 0 until cols) {
                val u = col.toFloat() / (cols - 1)
                val v = row.toFloat() / (rows - 1)

                // NDC: x ∈ [-1, 1], y ∈ [-1, 1]
                val x = u * 2f - 1f
                val y = v * 2f - 1f

                positions.put(x)
                positions.put(y)

                uvs.put(u)
                uvs.put(v)

                // Sample depth from the map.
                val depth = sampleDepth(u, v)
                depths.put(depth)
            }
        }
    }

    /**
     * Samples the depth map at normalised coordinates (u, v) using
     * nearest-neighbour lookup. 16-bit values are normalised to [0, 1].
     *
     * Foreground vertices (> [FOREGROUND_THRESHOLD]) get 1.5× amplification.
     */
    private fun sampleDepth(u: Float, v: Float): Float {
        val px = (u * (imageWidth - 1)).toInt().coerceIn(0, imageWidth - 1)
        val py = (v * (imageHeight - 1)).toInt().coerceIn(0, imageHeight - 1)

        val index = (py * imageWidth + px) * 2  // 16-bit per pixel

        if (index + 1 >= depthBytes.size) return 0f

        // Little-endian unsigned 16-bit.
        val lo = depthBytes[index].toInt() and 0xFF
        val hi = depthBytes[index + 1].toInt() and 0xFF
        val raw = (hi shl 8) or lo
        var normalised = raw / 65535f

        // Foreground amplification.
        if (normalised > FOREGROUND_THRESHOLD) {
            normalised *= FOREGROUND_AMPLIFICATION
            normalised = normalised.coerceAtMost(1f)
        }

        return normalised
    }

    // ─── Index generation ─────────────────────────────────────────────────

    private fun buildIndices(cols: Int, rows: Int) {
        for (row in 0 until rows - 1) {
            for (col in 0 until cols - 1) {
                val topLeft = row * cols + col
                val topRight = topLeft + 1
                val bottomLeft = (row + 1) * cols + col
                val bottomRight = bottomLeft + 1

                // First triangle.
                indices.put(topLeft)
                indices.put(bottomLeft)
                indices.put(topRight)

                // Second triangle.
                indices.put(topRight)
                indices.put(bottomLeft)
                indices.put(bottomRight)
            }
        }
    }

    // ─── Helpers ──────────────────────────────────────────────────────────

    companion object {
        /** Depth threshold above which a vertex is considered foreground. */
        const val FOREGROUND_THRESHOLD = 0.55f

        /** Z amplification factor for foreground vertices. */
        const val FOREGROUND_AMPLIFICATION = 1.5f

        private fun allocFloat(count: Int): FloatBuffer {
            return ByteBuffer.allocateDirect(count * 4)
                .order(ByteOrder.nativeOrder())
                .asFloatBuffer()
        }

        private fun allocInt(count: Int): IntBuffer {
            return ByteBuffer.allocateDirect(count * 4)
                .order(ByteOrder.nativeOrder())
                .asIntBuffer()
        }
    }
}

File: ios/Classes/DepthMesh.swift
import Foundation

/// Builds a planar triangle mesh displaced by a depth map for Metal rendering.
///
/// The mesh is a regular grid of `resolution × resolution` quads, each split
/// into two triangles. Positions are in NDC [-1, 1], UVs are [0, 1], and
/// per-vertex depth values are sampled from the supplied 16-bit depth map.
///
/// Foreground vertices (depth > 0.55) receive 1.5× amplified Z displacement.
class DepthMesh {

    /// Foreground / background threshold.
    static let foregroundThreshold: Float = 0.55

    /// Z amplification for foreground vertices.
    static let foregroundAmplification: Float = 1.5

    /// XY positions in NDC, 2 floats per vertex.
    let positions: [Float]

    /// UV texture coordinates, 2 floats per vertex.
    let uvs: [Float]

    /// Normalised depth values, 1 float per vertex.
    let depths: [Float]

    /// Triangle indices (UInt32).
    let indices: [UInt32]

    /// Total number of vertices.
    let vertexCount: Int

    /// Total number of index elements.
    let indexCount: Int

    /// Creates a new depth mesh.
    ///
    /// - Parameters:
    ///   - resolution: Grid subdivisions along each axis (16–256).
    ///   - imageWidth: Original image width in pixels.
    ///   - imageHeight: Original image height in pixels.
    ///   - depthData: Raw 16-bit unsigned LE depth map bytes.
    init(resolution: Int, imageWidth: Int, imageHeight: Int, depthData: Data) {
        let cols = resolution + 1
        let rows = resolution + 1
        vertexCount = cols * rows
        indexCount = resolution * resolution * 6

        var pos = [Float]()
        pos.reserveCapacity(vertexCount * 2)

        var uv = [Float]()
        uv.reserveCapacity(vertexCount * 2)

        var dep = [Float]()
        dep.reserveCapacity(vertexCount)

        var idx = [UInt32]()
        idx.reserveCapacity(indexCount)

        let depthBytes = [UInt8](depthData)

        // ── Build vertices ──────────────────────────────────────────

        for row in 0..<rows {
            for col in 0..<cols {
                let u = Float(col) / Float(cols - 1)
                let v = Float(row) / Float(rows - 1)

                // NDC: [-1, 1]
                let x = u * 2.0 - 1.0
                let y = v * 2.0 - 1.0

                pos.append(x)
                pos.append(y)

                uv.append(u)
                uv.append(v)

                // Sample depth.
                let px = min(Int(u * Float(imageWidth - 1)), imageWidth - 1)
                let py = min(Int(v * Float(imageHeight - 1)), imageHeight - 1)
                let byteIndex = (py * imageWidth + px) * 2

                var depth: Float = 0
                if byteIndex + 1 < depthBytes.count {
                    let lo = UInt16(depthBytes[byteIndex])
                    let hi = UInt16(depthBytes[byteIndex + 1])
                    let raw = (hi << 8) | lo
                    depth = Float(raw) / 65535.0
                }

                // Foreground amplification.
                if depth > DepthMesh.foregroundThreshold {
                    depth = min(depth * DepthMesh.foregroundAmplification, 1.0)
                }

                dep.append(depth)
            }
        }

        // ── Build indices ───────────────────────────────────────────

        for row in 0..<(rows - 1) {
            for col in 0..<(cols - 1) {
                let topLeft = UInt32(row * cols + col)
                let topRight = topLeft + 1
                let bottomLeft = UInt32((row + 1) * cols + col)
                let bottomRight = bottomLeft + 1

                // First triangle.
                idx.append(topLeft)
                idx.append(bottomLeft)
                idx.append(topRight)

                // Second triangle.
                idx.append(topRight)
                idx.append(bottomLeft)
                idx.append(bottomRight)
            }
        }

        self.positions = pos
        self.uvs = uv
        self.depths = dep
        self.indices = idx
    }
}
